packages/hdwallet-provider/src/index.ts

```diff
@@ -182,6 +191,18 @@ class HDWalletProvider {
         },
         signPersonalMessage(...args: any[]) {
           this.signMessage(...args);
+        },
+        signTypedMessage({ data, from }: any, cb: any) {
+          const dataIfExists = data;
+          if (!dataIfExists) {
+            cb("No data to sign");
+          }
+          if (!tmpWallets[from]) {
+            cb("Account not found");
+          }
+          const pkey = tmpWallets[from].getPrivateKey();
+          const sig = signTypedData(pkey, { data });
```

Comment:
I think we might need to deserialize `data` that we pass to `signTypedData`. I think it will come in as a stringified JSON object, here is a code example I found https://github.com/weijiekoh/eip712-signing-demo/blob/master/sign.js#L76-L94 which stringifies the data param.

Looks like eth-sig-util expects an argument of type [MsgParams<TypedMessage>](https://github.com/MetaMask/eth-sig-util/blob/d890e312fa84f4876fef7f9063a05c1b629e674f/src/index.ts#L585) and not a string.

@dmihal provided an example to test against [here](https://github.com/trufflesuite/truffle/issues/3585) and his test runs successfully if I do a `JSON.parse(data)` on the receiving side here in `signTypedMessage`. Can you make this change?

Reply:
Nevermind, I think this method should be receiving the data as an object as they will be calling the method directly and not using an http request. EIP [here](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#specification-of-the-eth_signtypeddata-json-rpc)